### PR TITLE
[SM-1256] Add BulkSecretAuthorizationHandler

### PR DESCRIPTION
**Pull request by Thomas-Avery**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/SM-1256

## 📔 Objective

The purpose of this PR is to add a `BulkSecretAuthorizationHandler` and use it for the `GetSecretsByIdsAsync` endpoint.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

**Review by coltonhurst — COMMENTED** [review]
One small comment

File: bitwarden_license/src/Commercial.Infrastructure.EntityFramework/SecretsManager/Repositories/SecretRepository.cs [hunk 1]
```diff
@@ -434,4 +428,32 @@ await dbContext.ServiceAccount
...
+                s.Id,
+                s.Projects.Any(p =>
+                    p.ServiceAccountAccessPolicies.Any(ap => ap.ServiceAccountId == userId && ap.Read)),
+                s.Projects.Any(p =>
+                    p.ServiceAccountAccessPolicies.Any(ap => ap.ServiceAccountId == userId && ap.Write))
+            )),
+            _ => secrets.Select(s => new SecretAccess(s.Id, false, false))
+        };
+
+    private record SecretAccess(Guid Id, bool Read, bool Write);
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Infrastructure.EntityFramework/SecretsManager/Repositories/SecretRepository.cs:458]: There is no technical reason. 

I'm following the style guide listed in our contributing guide here https://contributing.bitwarden.com/contributing/code-style/csharp. 

Which points to https://stackoverflow.com/questions/150479/order-of-items-in-classes-fields-properties-constructors-methods/310967#310967 
Which then points to https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1201.md

There StyleCop calls for within a class, records should go last.

[the remark this answers is not in the document]
Comment (coltonhurst) [reply bitwarden_license/src/Commercial.Infrastructure.EntityFramework/SecretsManager/Repositories/SecretRepository.cs:458]: Fair enough!

**Review by coltonhurst — APPROVED** [review]

**Review by coltonhurst — APPROVED** [review]

File: src/Api/SecretsManager/Controllers/SecretsController.cs [hunk 2]
```diff
@@ -260,25 +260,13 @@ public async Task<ListResponseModel<BulkDeleteResponseModel>> BulkDeleteAsync([F
             throw new NotFoundException();
         }
 
-        // Ensure all secrets belong to the same organization.
-        var organizationId = secrets.First().OrganizationId;
-        if (secrets.Any(secret => secret.OrganizationId != organizationId) ||
-            !_currentContext.AccessSecretsManager(organizationId))
+        var authorizationResult = await _authorizationService.AuthorizeAsync(User, secrets, BulkSecretOperations.ReadAll);
```
Comment (mzieniukbw) [inline src/Api/SecretsManager/Controllers/SecretsController.cs:263]: could you add integration test coverage for this ?
we already have two, but i think we should also have the edge cases, specifically:

1. First condition (line 258):
    1. no secret id provided (empty list), covers `secrets.Any()`
    2. secret id's provided but not all found, covers `secrets.Count != request.Ids.Count()`
2. Second condition (line 264) - Authorization did not succeed due to:
    1. Secrets not in the same organization
    2. Secret to which user does not have Read access

Also the success test only checks for the count, but not for content in the response, which is not good.

**Review by mzieniukbw — DISMISSED** [review]
LGTM.
Although the integration test coverage is kind of low.

File: src/Api/SecretsManager/Controllers/SecretsController.cs (hunk 2 above)
Comment (Thomas-Avery) [reply src/Api/SecretsManager/Controllers/SecretsController.cs:263]: Added in https://github.com/bitwarden/server/pull/4099/commits/76f3e9ad4e950908bfadce0e28d20f5cc2de0efc

**Review by mzieniukbw — APPROVED** [review]
